### PR TITLE
Use Vec::from_iter in Buffer::from_iter

### LINKED ISSUE #4937: Consider Vec::from_iter in Buffer::from_iter
### The source of unsoundness
Hi, we are the Rust researchers from [SunLab](https://sunlab-gmu.github.io/). Recently, we are trying to test our static analysis tool and find the potential unsound function in the package of `arrow-buffer`:
https://github.com/apache/arrow-rs/blob/90bc5ec96b5ae5162f469f9784dde7b1a53a5bdd/arrow-buffer/src/buffer/immutable.rs#L417-L432
In line 429, when the mutable pointer is cast to generic type, it is actually casting from `u8` to any-aligned type which could create a misaligned pointer. However, it seems that the programmers are also aware of the bugs in the test case:
https://github.com/apache/arrow-rs/blob/90bc5ec96b5ae5162f469f9784dde7b1a53a5bdd/arrow-buffer/src/buffer/scalar.rs#L224-L231
Here, `u8` is cast to `i32` and create a misaligned pointer and panic occurs.  
We consider that it should not happen in safe API anyway based on the [safety requirement](https://doc.rust-lang.org/std/ptr/fn.write.html) of `ptr.write()`. The problem could be fixed by passing the layout of `element` to allocate the memory.
We would love to have more discussion, and please correct if we are wrong.

### PR DESCRIPTION
**Pull request by Kikkon**
# Which issue does this PR close?

Closes #4937 .

# Rationale for this change
 

# What changes are included in this PR?

Replace unsafe code with Vec::from_iter in Buffer::from_iter for improved safety and performance

# Are there any user-facing changes?

no

### CODE REVIEW

File: arrow-buffer/src/buffer/immutable.rs [hunk 1]
```diff
@@ -423,24 +423,11 @@ impl Buffer {
...
-                unsafe {
-                    std::ptr::write(buffer.as_mut_ptr() as *mut T, element);
-                    buffer.set_len(size);
-                }
-                buffer
-            }
-        };
-
-        buffer.extend_from_iter(iterator);
+        let vec = Vec::from_iter(iter);
```
Comment (tustvold) [inline arrow-buffer/src/buffer/immutable.rs:428]: This method will now copy the data twice

File: arrow-buffer/src/buffer/immutable.rs [hunk 2]
```diff
@@ -423,24 +423,11 @@ impl Buffer {
...
+        let mut buffer = MutableBuffer::new(vec.len() * size);
```
Comment (tustvold) [inline arrow-buffer/src/buffer/immutable.rs:429]: You could use Buffer::from_vec here to avoid a copy

**Comment by tustvold** [comment]
Have you run any benchmarks to confirm this doesn't regress performance. The arithmetic kernels might be a good candidate

**Comment by Kikkon** [comment]
@tustvold Thanks for your advice, I will first change it to the draft status. I will reopen it after I have finished supplementing the benchmark.

**Comment by Kikkon** [comment]
Hi @tustvold ,
After comparing with the baseline on the master branch, it appears that this change has no impact on performance. If there are no other concerns, perhaps it's ready for merging. By the way, are there any other issues suitable for newcomers? I think I might not be very familiar with this area yet.  (:

**Review by tustvold — APPROVED** [review]
I did a scan of this and it turns out we don't use this code for anything other than tests anymore, instead preferring either the trusted length or unary/binary methods

**Pull request merged**